x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java

```diff
@@ -78,16 +78,24 @@ public void startDeployment(TrainedModelDeploymentTask task, ActionListener<Trai
         doStartDeployment(task, listener);
     }
 
-    private void doStartDeployment(TrainedModelDeploymentTask task, ActionListener<TrainedModelDeploymentTask> listener) {
+    private void doStartDeployment(TrainedModelDeploymentTask task, ActionListener<TrainedModelDeploymentTask> finalListener) {
         logger.debug("[{}] Starting model deployment", task.getModelId());
 
-        ProcessContext processContext = new ProcessContext(task.getModelId(), task.getIndex(), executorServiceForProcess);
+        ProcessContext processContext = new ProcessContext(task.getModelId(), task.getIndex(), executorServiceForProcess, task.getId());
 
         if (processContextByAllocation.putIfAbsent(task.getId(), processContext) != null) {
-            listener.onFailure(ExceptionsHelper.serverError("[{}] Could not create process as one already exists", task.getModelId()));
+            finalListener.onFailure(ExceptionsHelper.serverError("[{}] Could not create process as one already exists", task.getModelId()));
             return;
         }
 
+        ActionListener<TrainedModelDeploymentTask> listener = ActionListener.wrap(
+            finalListener::onResponse,
+            failure -> {
+                processContextByAllocation.remove(task.getId());
```

Comment:
If starting the deployment fails in any way, we should not keep the process context around. We should remove it.